File: Algorithms/Easy/MergeStringsAlternately.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        first, second = 0, 0
        result = []
        while first < len(word1) and second < len(word2):
            if first == second:
                result.append(word1[first])
                first += 1
            else:
                result.append(word2[second])
                second += 1

        if first < len(word1):
            result.extend(word1[first:])

        if second < len(word2):
            result.extend(word2[second:])

        return "".join(result)

File: Algorithms/Easy/1768.Merge-Strings-Alternately/MergeStringsAlternately.go
package easy

// Time complexity: O(n)
// Space complexity: O(1)
func mergeAlternately(word1 string, word2 string) string {
	m, n := len(word1), len(word2)
	res := make([]byte, 0, m+n)

	ptr1, ptr2 := 0, 0
	for ptr1 < m && ptr2 < n {
		if ptr1 == ptr2 {
			res = append(res, word1[ptr1])
			ptr1++
		} else {
			res = append(res, word2[ptr2])
			ptr2++
		}
	}

	if ptr1 < m {
		res = append(res, word1[ptr1:]...)
	}
	if ptr2 < n {
		res = append(res, word2[ptr2:]...)
	}

	return string(res)
}
